fv3core/stencils/dyn_core.py

```diff
@@ -179,381 +197,415 @@ def dyncore_temporaries(shape):
+            # they will be re-computed from scratch on the next acoustic timestep.
+
+            if self.do_halo_exchange:
+                for halovar in ["delp_quantity", "pt_quantity", "q_con_quantity"]:
+                    self.comm.halo_update(
+                        state.__getattribute__(halovar), n_points=utils.halo
+                    )
+
+            # Not used unless we implement other betas and alternatives to nh_p_grad
+            # if self.namelist.d_ext > 0:
+            #    raise 'Unimplemented namelist option d_ext > 0'
+            # else:
+            #    divg2 = utils.make_storage_from_shape(delz.shape,
+            #   grid.compute_origin())
 
-        if global_config.get_do_halo_exchange():
-            if it != n_split - 1:
-                reqs_vector = comm.start_vector_halo_update(
-                    state.u_quantity, state.v_quantity, n_points=utils.halo
+            if not hydrostatic:
+                updatedzd.compute(
+                    state.dp_ref,
+                    state.zs,
+                    state.zh,
+                    state.crx,
+                    state.cry,
+                    state.xfx,
+                    state.yfx,
+                    state.wsd,
+                    dt,
                 )
-            else:
-                if spec.namelist.grid_type < 4:
-                    comm.synchronize_vector_interfaces(
-                        state.u_quantity, state.v_quantity
+
+                riem_solver3.compute(
+                    remap_step,
+                    dt,
+                    akap,
+                    state.cappa,
+                    state.ptop,
+                    state.zs,
+                    state.w,
+                    state.delz,
+                    state.q_con,
+                    state.delp,
+                    state.pt,
+                    state.zh,
+                    state.pe,
+                    state.pkc,
+                    state.pk3,
+                    state.pk,
+                    state.peln,
+                    state.wsd,
+                )
+
+                if self.do_halo_exchange:
+                    reqs["zh_quantity"] = self.comm.start_halo_update(
+                        state.zh_quantity, n_points=utils.halo
                     )
+                    if grid.npx == grid.npy:
+                        reqs["pkc_quantity"] = self.comm.start_halo_update(
+                            state.pkc_quantity, n_points=2
+                        )
+                    else:
+                        reqs["pkc_quantity"] = self.comm.start_halo_update(
+                            state.pkc_quantity, n_points=utils.halo
+                        )
+                if remap_step:
+                    pe_halo.compute(state.pe, state.delp, state.ptop)
+                if self.namelist.use_logp:
+                    raise Exception("unimplemented namelist option use_logp=True")
```

Comment:
```suggestion
                    raise NotImplementedError("unimplemented namelist option use_logp=True")
```
The base class `Exception` should never be raised, since it can only ever be caught with `except Exception:`, which catches every exception, including things like system errors you shouldn't catch. Some people go as far as to say you should never raise base exception types so that someone can always catch your particular exception, which is debatable. At least it makes sense to use `NotImplementedError` over `Exception` for features which aren't implemented.

Reply:
never say never :). I agree it's good to be specific with error types and I'll make the change here. I disagree that one should never raise Exception. You may want to have an exception that can't be caught, you may not have the time to go with a better one.